File: services/api/src/api/routers/attributes.py
"""Document attribute definition routes."""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from api.auth.dependencies import OrgContext, require_org_admin
from api.dependencies import get_tenant_db
from api.repositories.attribute import AttributeDefinitionRepository
from api.schemas.attribute import (
    AttributeDefinitionCreate,
    AttributeDefinitionRead,
    AttributeDefinitionUpdate,
)

router = APIRouter()


@router.get("/", response_model=list[AttributeDefinitionRead])
async def list_attributes(
    ctx: Annotated[OrgContext, Depends(require_org_admin)],
    session: Annotated[AsyncSession, Depends(get_tenant_db)],
) -> list[AttributeDefinitionRead]:
    repo = AttributeDefinitionRepository(session)
    items = await repo.list_all()
    return [AttributeDefinitionRead.model_validate(i) for i in items]


@router.get("/{attribute_id}", response_model=AttributeDefinitionRead)
async def get_attribute(
    attribute_id: uuid.UUID,
    ctx: Annotated[OrgContext, Depends(require_org_admin)],
    session: Annotated[AsyncSession, Depends(get_tenant_db)],
) -> AttributeDefinitionRead:
    repo = AttributeDefinitionRepository(session)
    instance = await repo.get(attribute_id)
    if instance is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Not found")
    return AttributeDefinitionRead.model_validate(instance)


@router.post(
    "/", response_model=AttributeDefinitionRead, status_code=status.HTTP_201_CREATED
)
async def create_attribute(
    body: AttributeDefinitionCreate,
    ctx: Annotated[OrgContext, Depends(require_org_admin)],
    session: Annotated[AsyncSession, Depends(get_tenant_db)],
) -> AttributeDefinitionRead:
    repo = AttributeDefinitionRepository(session)
    instance = await repo.create(
        name=body.name,
        slug=body.slug,
        org_id=ctx.org_id,
        attribute_type=body.attribute_type,
        picklist_options=body.picklist_options,
        required=body.required,
        order=body.order,
    )
    return AttributeDefinitionRead.model_validate(instance)


@router.patch("/{attribute_id}", response_model=AttributeDefinitionRead)
async def update_attribute(
    attribute_id: uuid.UUID,
    body: AttributeDefinitionUpdate,
    ctx: Annotated[OrgContext, Depends(require_org_admin)],
    session: Annotated[AsyncSession, Depends(get_tenant_db)],
) -> AttributeDefinitionRead:
    repo = AttributeDefinitionRepository(session)
    instance = await repo.get(attribute_id)
    if instance is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Not found")

    instance = await repo.update(
        instance,
        name=body.name,
        attribute_type=body.attribute_type,
        picklist_options=body.picklist_options,
        required=body.required,
        order=body.order,
    )
    return AttributeDefinitionRead.model_validate(instance)


@router.delete("/{attribute_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_attribute(
    attribute_id: uuid.UUID,
    ctx: Annotated[OrgContext, Depends(require_org_admin)],
    session: Annotated[AsyncSession, Depends(get_tenant_db)],
) -> None:
    repo = AttributeDefinitionRepository(session)
    deleted = await repo.delete(attribute_id)
    if not deleted:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Not found")

File: ui/src/lib/api/attributes.ts
import apiClient from "./client";

export type AttributeType = "freeform" | "picklist";

export interface AttributeDefinition {
  id: string;
  name: string;
  slug: string;
  attribute_type: AttributeType;
  picklist_options: string[];
  required: boolean;
  order: number;
}

export interface AttributeCreateInput {
  name: string;
  slug: string;
  attribute_type: AttributeType;
  picklist_options?: string[];
  required?: boolean;
  order?: number;
}

export interface AttributeUpdateInput {
  name?: string;
  attribute_type?: AttributeType;
  picklist_options?: string[];
  required?: boolean;
  order?: number;
}

export async function listAttributes(): Promise<AttributeDefinition[]> {
  const response = await apiClient.get<{ items: AttributeDefinition[] }>("/attributes/", {
    params: { page_size: 200 },
  });
  return response.data.items;
}

export async function createAttribute(
  input: AttributeCreateInput,
): Promise<AttributeDefinition> {
  const response = await apiClient.post<AttributeDefinition>("/attributes/", input);
  return response.data;
}

export async function updateAttribute(
  id: string,
  input: AttributeUpdateInput,
): Promise<AttributeDefinition> {
  const response = await apiClient.patch<AttributeDefinition>(`/attributes/${id}`, input);
  return response.data;
}

export async function deleteAttribute(id: string): Promise<void> {
  await apiClient.delete(`/attributes/${id}`);
}
